subsys/bluetooth/mesh/adv.c

```diff
@@ -84,16 +88,66 @@ struct net_buf *bt_mesh_adv_create_from_pool(struct net_buf_pool *pool,
 	(void)memset(adv, 0, sizeof(*adv));
 
 	adv->type         = type;
+	adv->tag          = tag;
 	adv->xmit         = xmit;
 
 	return buf;
 }
 
-struct net_buf *bt_mesh_adv_create(enum bt_mesh_adv_type type, uint8_t xmit,
-				   k_timeout_t timeout)
+struct net_buf *bt_mesh_adv_create(enum bt_mesh_adv_type type,
+				   enum bt_mesh_adv_tag tag,
+				   uint8_t xmit, k_timeout_t timeout)
 {
 	return bt_mesh_adv_create_from_pool(&adv_buf_pool, adv_alloc, type,
-					    xmit, timeout);
+					    tag, xmit, timeout);
+}
+
+static struct net_buf *process_events(struct k_poll_event *ev, int count)
+{
+	for (; count; ev++, count--) {
+		BT_DBG("ev->state %u", ev->state);
+
+		switch (ev->state) {
+		case K_POLL_STATE_FIFO_DATA_AVAILABLE:
+			return net_buf_get(ev->fifo, K_NO_WAIT);
+		case K_POLL_STATE_NOT_READY:
+		case K_POLL_STATE_CANCELLED:
+			break;
+		default:
+			BT_WARN("Unexpected k_poll event state %u", ev->state);
+			break;
+		}
+	}
+
+	return NULL;
+}
+
+struct net_buf *bt_mesh_adv_buf_get(k_timeout_t timeout)
+{
+	int err;
+	static struct k_poll_event events[] = {
+		K_POLL_EVENT_STATIC_INITIALIZER(K_POLL_TYPE_FIFO_DATA_AVAILABLE,
+						K_POLL_MODE_NOTIFY_ONLY,
+						&bt_mesh_adv_queue,
+						0),
+#if defined(CONFIG_BT_MESH_RELAY)
+		K_POLL_EVENT_STATIC_INITIALIZER(K_POLL_TYPE_FIFO_DATA_AVAILABLE,
+						K_POLL_MODE_NOTIFY_ONLY,
+						&bt_mesh_relay_queue,
+						0),
+#endif /* CONFIG_BT_MESH_RELAY */
+	};
+
+	for (int i = 0; i < ARRAY_SIZE(events); i++) {
+		events[i].state = K_POLL_STATE_NOT_READY;
+	}
```

Comment:
Thanks for the explanation on this. Would it make sense to put the poll events on stack instead, and initialize them with `k_poll_event_init`? Should save 80 byte of static ram in adv_ext mode, since that won't max out the stack depth

Reply:
Looks Make sense to me